Sources/ContentScrollView.swift

```diff
@@ -83,40 +94,81 @@ open class ContentScrollView: UIScrollView {
 
         for i in 0...currentIndex {
-            guard let pageView = dataSource.contentScrollView(self, viewForPageAt: i) else { return }
-            pageViews.append(pageView)
-            addSubview(pageView)
+            let containerView = UIView()
+            pageViews.append(containerView)
+            addSubview(containerView)
 
             let leadingAnchor = i > 0 ? pageViews[i - 1].trailingAnchor : self.leadingAnchor
-            pageView.translatesAutoresizingMaskIntoConstraints = false
+            containerView.translatesAutoresizingMaskIntoConstraints = false
             NSLayoutConstraint.activate([
-                pageView.topAnchor.constraint(equalTo: self.topAnchor),
-                pageView.widthAnchor.constraint(equalTo: self.widthAnchor),
-                pageView.heightAnchor.constraint(equalTo: self.heightAnchor),
-                pageView.leadingAnchor.constraint(equalTo: leadingAnchor)
+                containerView.topAnchor.constraint(equalTo: self.topAnchor),
+                containerView.widthAnchor.constraint(equalTo: self.widthAnchor),
+                containerView.heightAnchor.constraint(equalTo: self.heightAnchor),
+                containerView.leadingAnchor.constraint(equalTo: leadingAnchor)
             ])
         }
 
         guard currentIndex < dataSource.numberOfPages(in: self) else { return }
         for i in (currentIndex + 1)..<dataSource.numberOfPages(in: self) {
-            guard let pageView = dataSource.contentScrollView(self, viewForPageAt: i) else { return }
-            pageViews.append(pageView)
-            addSubview(pageView)
+            let containerView = UIView()
+            pageViews.append(containerView)
+            addSubview(containerView)
+
+            containerView.translatesAutoresizingMaskIntoConstraints = false
+            NSLayoutConstraint.activate([
+                containerView.topAnchor.constraint(equalTo: self.topAnchor),
+                containerView.widthAnchor.constraint(equalTo: self.widthAnchor),
+                containerView.heightAnchor.constraint(equalTo: self.heightAnchor),
+                containerView.leadingAnchor.constraint(equalTo: pageViews[i - 1].trailingAnchor)
+            ])
+        }
+    }
+
+    /// lazy setup
+    /// - set current pageView and the buffered previous/next pageView
+    /// - remove outside pageView
+    private func setVisiblePageViews() {
+        guard let dataSource = self.dataSource else { return }
+        let previousPageIndex: Int = {
+            let i = currentIndex - previousBufferPageCount
+            return (i > 0) ? i : 0
+        }()
+        let nextPageIndex: Int = {
+            let i = currentIndex + nextBufferPageCount
+            let lastIndex = dataSource.numberOfPages(in: self) - 1
+            return (i > lastIndex) ? lastIndex : i
+        }()
+        let range = previousPageIndex...nextPageIndex
```

Comment:
This is unsafe code which causes crash. Maybe it worth adding check for valid range.
```swift
if previousPageIndex > nextPageIndex {
    return
}
```